### PR TITLE
Implement more of InGameUI

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/hooker/setuphooks_zh.cpp [hunk 1]
```diff
@@ -3771,6 +3773,13 @@ void Setup_Hooks()
 
     // w3dingameui.h
     Hook_Any(0x007A9B30, W3DInGameUI::Create_View);
+    Hook_Any(0x007A95A0, W3DInGameUI::Init);
+    Hook_Any(0x007A95C0, W3DInGameUI::Reset);
+    Hook_Any(0x007A95B0, W3DInGameUI::Update);
+    Hook_Any(0x007A95D0, W3DInGameUI::Draw);
+    Hook_Any(0x007A9650, W3DInGameUI::Draw_Selection_Region);
+    // Hook_Any(0x007A9690, W3DInGameUI::Draw_Move_Hints);
```
Comment (xezon) [inline src/hooker/setuphooks_zh.cpp]: Is this commented on purpose?

File: src/hooker/setuphooks_zh.cpp [hunk 2]
```diff
@@ -1,4 +1,5 @@
 ﻿/**
+/**
```
Comment (xezon) [inline src/hooker/setuphooks_zh.cpp]: Can remove.

File: src/platform/w3dengine/client/w3dingameui.cpp [hunk 3]
```diff
@@ -13,9 +13,225 @@
...
+}
+
 View *W3DInGameUI::Create_View()
 {
     return new W3DView();
 }
+
+void W3DInGameUI::Draw_Selection_Region()
+{
+    g_theDisplay->Draw_Open_Rect(m_dragSelectRegion.lo.x,
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dingameui.cpp:88]:
```c++
virtual void Draw_Open_Rect(int, int, int, int, float, unsigned) = 0;
```
to
```c++
virtual void Draw_Open_Rect(int x, int y, int w, int h, float line, unsigned int color) = 0;
```

Or use the names from

```c++
void W3DDisplay::Draw_Open_Rect(int32_t x, int32_t y, int32_t width, int32_t height, float border_width, uint32_t color)
```

File: src/platform/w3dengine/client/w3dingameui.cpp [hunk 4]
```diff
@@ -13,9 +13,225 @@
...
+    Matrix3D tm;
+
+    for (;;) {
+        if ((g_theGameClient->Get_Frame() - m_moveHint[count].frame) > 40) {
+            if (m_moveHintRenderObj[count] != nullptr && !m_moveHintRenderObj[count]->Is_Hidden()) {
+                m_moveHintRenderObj[count]->Set_Hidden(true);
+                W3DDisplay::s_3DScene->Remove_Render_Object(m_moveHintRenderObj[count]);
+            }
+
+            goto label1;
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dingameui.cpp]: Can remove `goto` by writing here

```c++
if (++count >= MAX_MOVE_HINTS) {
    return;
}
continue;
```

File: src/platform/w3dengine/client/w3dingameui.cpp [hunk 5]
```diff
@@ -13,9 +13,225 @@
...
+        goto label2;
+    }
+}
+
+void W3DInGameUI::Draw_Attack_Hints(View *view) {}
+
+void W3DInGameUI::Draw_Place_Angle(View *view)
+{
+    if (m_locator01 == nullptr) {
+        m_locator01 = W3DDisplay::s_assetManager->Create_Render_Obj("Locater01");
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dingameui.cpp]: Not `"Locator01"` ?

File: src/platform/w3dengine/client/w3dingameui.cpp [hunk 6]
```diff
@@ -13,9 +13,225 @@
...
+
+        if (m_locator01 == nullptr) {
+            return;
+        }
+    }
+
+    if (m_locator02 == nullptr) {
+        m_locator02 = W3DDisplay::s_assetManager->Create_Render_Obj("Locater02");
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dingameui.cpp]: Not `"Locator02"` ?

File: src/game/client/ingameui.cpp [hunk 7]
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
     }
 }
 
+InGameUI::InGameUI() :
```
Comment (xezon) [inline src/game/client/ingameui.cpp:261]: `m_dragSelectRegion`
`m_scrollAmt`
`m_isQuitMenuVisible`
`m_cameraRotateLeft`
`m_cameraRotateRight`
`m_cameraZoomIn`
`m_cameraDrawingTrackable`
`m_cameraZoomOut`
`m_noRadarEdgeSound`

are not initialized

File: src/game/client/ingameui.cpp [hunk 8]
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
...
+        m_moveHint[i].frame = 0;
+    }
+
+    for (int i = 0; i < MAX_BUILD_PROGRESS; i++) {
+        m_buildProgress[i].m_thingTemplate = nullptr;
+        m_buildProgress[i].m_percentComplete = 0.0f;
+        m_buildProgress[i].m_control = nullptr;
+    }
+
+    m_placeIcon = new Drawable *[g_theWriteableGlobalData->m_maxLineBuildObjects];
```
Comment (xezon) [inline src/game/client/ingameui.cpp:373]: Can be optimized to

```c++
m_placeIcon = new Drawable *[g_theWriteableGlobalData->m_maxLineBuildObjects]();
```

for 0 initialization.

File: src/game/client/ingameui.cpp [hunk 9]
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
...
+    m_namedTimerFlashDuration(0.0f),
+    m_namedTimerLastFlashFrame(0),
+    m_namedTimerUsedFlashColor(true),
+    m_showNamedTimers(true),
+    m_namedTimerNormalFont("Arial"),
+    m_namedTimerNormalPointSize(10),
+    m_namedTimerNormalBold(false),
+    m_namedTimerReadyFont("Arial"),
+    m_namedTimerReadyPointSize(10),
+    m_namedTimerReadyBold(0),
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: false

File: src/game/client/ingameui.cpp [hunk 10]
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
...
+    m_drawableCaptionFont("Arial"),
+    m_drawableCaptionPointSize(10),
+    m_drawableCaptionBold(false),
+    m_tooltipsDisabled(false),
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: 0 ?
Type is put as unsigned int.

File: src/game/client/ingameui.cpp [hunk 11]
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
...
+        m_placeIcon[i] = nullptr;
+    }
+
+    m_placeAnchorStart.x = 0;
+    m_placeAnchorStart.y = 0;
+    m_placeAnchorEnd.x = 0;
+    m_placeAnchorEnd.y = 0;
+
+    for (int i = 0; i < MAX_UI_MESSAGES; i++) {
+        m_uiMessages[i].full_text.Clear();
```
Comment (xezon) [inline src/game/client/ingameui.cpp:385]: Is obsolete call.

File: src/game/client/ingameui.cpp [hunk 12]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+void InGameUI::Play_Cameo_Movie(const Utf8String &movie_name)
+{
+    Stop_Cameo_Movie();
+    m_cameoVideoStream = g_theVideoPlayer->Open(movie_name);
+
+    if (m_cameoVideoStream != nullptr) {
+        m_currentlyPlayingMovie = movie_name;
+        m_cameoVideoBuffer = g_theDisplay->Create_VideoBuffer();
+
+        if (m_cameoVideoBuffer == nullptr
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: != nullptr

File: src/game/client/ingameui.cpp [hunk 13]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+        }
+    }
+}
+
+void InGameUI::Stop_Cameo_Movie()
+{
+    GameWindow *hud =
+        g_theWindowManager->Win_Get_Window_From_Id(nullptr, g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:RightHUD"));
+    hud->Win_Get_Instance_Data()->Set_VideoBuffer(nullptr);
+    if (m_cameoVideoBuffer != nullptr) {
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3231]: I suggest add blank line above this.

File: src/game/client/ingameui.cpp [hunk 14]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+}
+
+int InGameUI::Select_Matching_Across_Map()
+{
+    int ret = Select_Matching_Across_Region(nullptr);
+
+    if (ret == -1) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:NothingSelected"));
+    } else if (ret != 0) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: GUI:SelectedAcrossMap ?

File: src/game/client/ingameui.cpp [hunk 15]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    } else {
+        Drawable *draw = g_theInGameUI->Get_First_Selected_Drawable();
+
+        if (draw == nullptr || draw->Get_Object() != nullptr || !draw->Get_Object()->Is_KindOf(KINDOF_STRUCTURE)) {
+            g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: GUI:SelectedAcrossMap ?

File: src/game/client/ingameui.cpp [hunk 16]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    const ThingTemplate *tmplate = data->tmplate;
+
+    if (draw != nullptr) {
+        Object *obj = draw->Get_Object();
+
+        if (obj == nullptr) {
+            return false;
+        }
+
+        bool equivalent = obj->Get_Template()->Is_Equivalent_To(tmplate);
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: is_equivalent

File: src/game/client/ingameui.cpp [hunk 17]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+}
+
+int InGameUI::Select_All_Units_By_Type_Across_Map(BitFlags<KINDOF_COUNT> must_be_set, BitFlags<KINDOF_COUNT> must_be_clear)
+{
+    int ret = Select_All_Units_By_Type_Across_Region(nullptr, must_be_set, must_be_clear);
+
+    if (ret == -1) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:NothingSelected"));
+    } else if (ret != 0) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: GUI:SelectedAcrossMap ?

File: src/game/client/ingameui.cpp [hunk 18]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    } else {
+        Drawable *draw = g_theInGameUI->Get_First_Selected_Drawable();
+
+        if (draw == nullptr || draw->Get_Object() != nullptr || !draw->Get_Object()->Is_KindOf(KINDOF_STRUCTURE)) {
+            g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: GUI:SelectedAcrossMap ?

File: src/game/client/ingameui.cpp [hunk 19]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+        for (auto it = templates.begin(); it != templates.end(); it++) {
+            data.tmplate = *it;
+            data.is_carbomb = is_carbomb;
+
+            if (region != nullptr) {
+                count += g_theTacticalView->Iterate_Drawables_In_Region(region, Similar_Unit_Selection, &data);
+            } else {
+                for (Drawable *drawable = g_theGameClient->First_Drawable(); drawable != nullptr;
+                     drawable = drawable->Get_Next()) {
+                    count += Similar_Unit_Selection(drawable, &data);
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3392]: This is adding bool value to count.

Perhaps meant to be

```c++
if (Similar_Unit_Selection(drawable, &data))
    count++;
```
?

File: src/game/client/ingameui.cpp [hunk 20]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+    return false;
+}
+
+int InGameUI::Select_All_Units_By_Type_Across_Region(
+    IRegion2D *region, BitFlags<KINDOF_COUNT> must_be_set, BitFlags<KINDOF_COUNT> must_be_clear)
+{
+    KindOfUnitSelectionData data;
+    int count = 0;
+    int size = Get_All_Selected_Drawables()->size();
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: previous_count

File: src/game/client/ingameui.cpp [hunk 21]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    } else {
+        region->lo.y = anchor->y;
+        region->hi.y = dest->y;
+    }
+}
+
+FloatingTextData::FloatingTextData() : m_color(0), m_frameTimeOut(0), m_frameCount(0)
+{
+    m_pos3D.Zero();
+    m_text.Clear();
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3576]: Is obsolete call.

File: src/game/client/ingameui.cpp [hunk 22]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+            GameMessage *msg = g_theMessageStream->Append_Message(GameMessage::MSG_CREATE_SELECTED_GROUP);
+            msg->Append_Bool_Arg(true);
+            msg->Append_ObjectID_Arg(obj->Get_ID());
+            Select_Drawable(obj->Get_Drawable());
+            g_theTacticalView->Look_At(obj->Get_Position());
+        }
+    }
+}
+
+void InGameUI::Disable_Tooltips_Until(unsigned int until)
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: frame

File: src/game/client/ingameui.cpp [hunk 23]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+void InGameUI::Show_Idle_Worker_Layout()
+{
+    if (m_idleWorkerWin != nullptr) {
+        m_idleWorkerWin->Win_Enable(true);
+        m_idleWorkerCount = Get_Idle_Worker_Count();
+    } else {
+        m_idleWorkerWin = g_theWindowManager->Win_Get_Window_From_Id(
+            nullptr, g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonIdleWorker"));
+        captainslog_dbgassert(
+            m_idleWorkerWin != nullptr, "InGameUI::Show_Idle_Worker_Layout could not find IdleWorker.wnd to load ");
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3737]: Looks like `m_idleWorkerCount = Get_Idle_Worker_Count();` is missing in this branch.

File: src/game/client/ingameui.cpp [hunk 24]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    for (auto it = m_idleWorkerLists[index].begin(); it != m_idleWorkerLists[index].end(); it++) {
+        if (*it == obj) {
+            return *it;
+        }
+    }
+
+    return nullptr;
+}
+
+void InGameUI::Show_Idle_Worker_Layout()
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3728]: No call to `Gadget_Button_Set_Text` in this function?

File: src/game/client/ingameui.cpp [hunk 25]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+{
+    Unregister_Window_Layout(layout);
+    m_windowLayoutList.push_back(layout);
+}
+
+void InGameUI::Unregister_Window_Layout(WindowLayout *layout)
+{
+    for (auto it = m_windowLayoutList.begin(); it != m_windowLayoutList.end(); it++) {
+        if (*it == layout) {
+            m_windowLayoutList.erase(it);
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3814]: `return`, `break` or `it =` missing.

File: src/game/client/ingameui.cpp [hunk 26]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+{
+    ViewLocation loc;
+    g_theTacticalView->Get_Location(&loc);
+    g_theTacticalView->Reset_Camera(loc.Get_Pos(), 1, 0.0f, 0.0f);
+}
+
+bool InGameUI::Can_Selected_Objects_Non_Attack_Interact_With_Object(
+    const Object *object_to_interact_with, SelectionRules rule)
+{
+    for (int i = ACTIONTYPE_ATTACK_OBJECT; i < NUM_ACTIONTYPES; i++) {
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3830]: This looks odd. Effectively this loop is equivalent to simplified:

```c++
    for (int i = ACTIONTYPE_ATTACK_OBJECT + 1; i < NUM_ACTIONTYPES; i++) {
        if (Can_Selected_Objects_Do_Action(static_cast<ActionType>(i), object_to_interact_with, rule, false)) {
            return true;
        }
    }
```

File: src/game/client/ingameui.cpp [hunk 27]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+CanAttackResult InGameUI::Get_Can_Selected_Objects_Attack(
+    ActionType action, const Object *object_to_interact_with, SelectionRules rule, bool force_to_attack) const
+{
+    if ((object_to_interact_with == nullptr && action != ACTIONTYPE_SET_RALLY_POINT)
+        || (object_to_interact_with != nullptr && action == ACTIONTYPE_SET_RALLY_POINT)) {
+        return ATTACK_RESULT_0;
+    }
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
+    int count = 0;
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3849]:
```

ATTACK_RESULT_0 = CANNOT_ATTACK
ATTACK_RESULT_1 = CAN_ATTACK_BUT_UNREACHABLE
ATTACK_RESULT_2 = CAN_ATTACK_BUT_OUT_OF_RANGE
ATTACK_RESULT_3 = CAN_ATTACK
```

File: src/game/client/ingameui.cpp [hunk 28]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    CanAttackResult result1 = ATTACK_RESULT_0;
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: any_can_attack

File: src/game/client/ingameui.cpp [hunk 29]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    CanAttackResult result2 = ATTACK_RESULT_3;
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: all_can_attack

File: src/game/client/ingameui.cpp [hunk 30]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+                    can_do = obj->Is_KindOf(KINDOF_AUTO_RALLYPOINT) && obj->Is_Locally_Controlled();
+                } else {
+                    can_do = false;
+                }
+
+                break;
+            }
+            case ACTIONTYPE_COMBAT_DROP:
+                can_do = g_theActionManager->Can_Enter_Object(
+                    draw->Get_Object(), object_to_interact_with, COMMANDSOURCE_PLAYER, CAN_ENTER_2);
```
Comment (xezon) [inline src/game/client/ingameui.cpp:3995]: CAN_ENTER_2 = CAN_ENTER_WITH_COMBAT_DROP

File: src/game/client/ingameui.cpp [hunk 31]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+    return rule == SELECTION_ALL && count > 0 && qualify == count;
+}
+
+bool InGameUI::Can_Selected_Objects_Do_Special_Power(CommandButton *command,
+    const Object *object_to_interact_with,
+    Coord3D *position,
+    SelectionRules rule,
+    unsigned int options,
+    Object *o) const
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: custom_object

File: src/game/client/ingameui.cpp [hunk 32]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+    Drawable *draw;
+
+    if (o != nullptr) {
+        draw = o->Get_Drawable();
+    } else {
+        draw = nullptr;
+    }
+
+    std::list<Drawable *> list;
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: custom_drawables

File: src/game/client/ingameui.cpp [hunk 33]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+    if (draw != nullptr) {
+        list.push_back(draw);
+    }
+
+    const std::list<Drawable *> *list2;
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: drawables

File: src/game/client/ingameui.cpp [hunk 34]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+    if (needs_target_object && object_to_interact_with == nullptr) {
+        return false;
+    }
+
+    if (needs_target_pos && position == nullptr) {
+        return false;
+    }
+
+    Drawable *draw;
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: custom_drawable

File: src/game/client/ingameui.cpp [hunk 35]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+    }
+
+    int count = 0;
+    int qualify = 0;
+
+    for (auto it = list2->begin(); it != list2->end(); it++) {
+        Drawable *drawable = *it;
+        count++;
+
+        if (needs_target_object || needs_target_pos) {
```
Comment (xezon) [inline src/game/client/ingameui.cpp:4067]: Overcomplicated condition tree.

Simpler:

```c++
if (needs_target_object) {
...
} else if (needs_target_pos) {
...
} else {
...
}
```

File: src/game/client/ingameui.cpp [hunk 36]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+                        true)) {
+                    if (rule == SELECTION_ANY) {
+                        return true;
+                    }
+
+                    qualify++;
+                }
+            }
+        } else {
+            if (g_theActionManager->Can_Do_Special_Power(draw->Get_Object(), power, COMMANDSOURCE_PLAYER, options, 1)) {
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: `drawable`, no ?
Otherwise this may be null.

File: src/game/client/ingameui.cpp [hunk 37]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+    if (needs_target_object && object_to_interact_with == nullptr) {
+        return false;
+    }
+
+    if (needs_target_pos && position == nullptr) {
+        return false;
+    }
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: drawables

File: src/game/client/ingameui.cpp [hunk 38]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+bool InGameUI::Can_Selected_Objects_Do_Action(
+    ActionType action, const Object *object_to_interact_with, SelectionRules rule, bool additional_checking) const
+{
+    if ((object_to_interact_with == nullptr && action != ACTIONTYPE_SET_RALLY_POINT)
+        || (object_to_interact_with != nullptr && action == ACTIONTYPE_SET_RALLY_POINT)) {
+        return ATTACK_RESULT_0;
+    }
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: drawables

File: src/game/client/ingameui.cpp [hunk 39]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+
+    return rule == SELECTION_ALL && count > 0 && qualify == count;
+}
+
+bool InGameUI::Can_Selected_Objects_Override_Special_Power_Destination(
+    const Coord3D *pos, SelectionRules rule, SpecialPowerType type) const
+{
+    int count = 0;
+    int qualify = 0;
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
```
Comment (xezon) [inline src/game/client/ingameui.cpp]: drawables

File: src/game/client/ingameui.cpp [hunk 40]
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
...
+        return false;
+    }
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
+    int count = 0;
+    int qualify = 0;
+
+    for (auto it = list->begin(); it != list->end(); it++) {
+        Drawable *draw = *it;
+        if (needs_target_object || needs_target_pos) {
```
Comment (xezon) [inline src/game/client/ingameui.cpp:4150]: Overcomplicated condition tree.

Simpler:

```c++
if (needs_target_object) {
...
} else if (needs_target_pos) {
...
} else {
...
}
```

File: src/platform/w3dengine/client/w3dingameui.cpp [hunk 41]
```diff
@@ -13,9 +13,225 @@
...
+            }
+
+            goto label1;
+        }
+
+        if (m_moveHintRenderObj[count] == nullptr) {
+            break;
+        }
+
+    label2:
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dingameui.cpp:119]: Is there no version of this function without this label?

File: src/hooker/setuphooks_zh.cpp (hunk 1 above)
Comment (jonwil) [reply src/hooker/setuphooks_zh.cpp]: Fixed.

File: src/hooker/setuphooks_zh.cpp (hunk 2 above)
Comment (jonwil) [reply src/hooker/setuphooks_zh.cpp]: fixed.

File: src/platform/w3dengine/client/w3dingameui.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dingameui.cpp]: fixed.

File: src/platform/w3dengine/client/w3dingameui.cpp (hunk 5 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dingameui.cpp]: Locater is correct spelling here, I renamed some variables.

File: src/platform/w3dengine/client/w3dingameui.cpp (hunk 6 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dingameui.cpp]: Locater is correct spelling here, I renamed some variables.

File: src/game/client/ingameui.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp:261]: fixed.

File: src/game/client/ingameui.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp:3231]: fixed.

File: src/game/client/ingameui.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp:3392]: Nope, original code does this.

File: src/game/client/ingameui.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 23 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp:3737]: Original doesn't have it.

File: src/game/client/ingameui.cpp (hunk 24 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp:3728]: Original doesn't have it.

File: src/game/client/ingameui.cpp (hunk 25 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp:3814]: fixed.

File: src/game/client/ingameui.cpp (hunk 27 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp:3849]: fixed.

File: src/game/client/ingameui.cpp (hunk 28 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 29 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 32 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 31 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 33 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 34 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 36 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 37 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 38 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/game/client/ingameui.cpp (hunk 39 above)
Comment (jonwil) [reply src/game/client/ingameui.cpp]: fixed.

File: src/platform/w3dengine/client/w3dingameui.cpp (hunk 41 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dingameui.cpp:119]: I can't find anything cleaner.

**Review by xezon — COMMENTED** [review]
Game crashes when pressing Idle Worker button:

```
>	thyme.dll!InGameUI::Select_Next_Idle_Worker() Line 3646	C++
 	thyme.dll!GameWindowManager::Win_Send_System_Msg(GameWindow * window, unsigned int msg, unsigned int data1, unsigned int data2) Line 451	C++
 	game.dat!005abdb2()	Unknown
 	game.dat![Frames below may be incorrect and/or missing, no symbols loaded for game.dat]	Unknown
 	thyme.dll!GameWindowManager::Win_Send_Input_Msg(GameWindow * window, unsigned int msg, unsigned int data1, unsigned int data2) Line 464	C++
 	thyme.dll!GameWindowManager::Win_Process_Mouse_Event(GameWindowMessage msg, ICoord2D * mouse_pos, void * data) Line 742	C++
 	game.dat!0061725f()	Unknown
 	thyme.dll!MessageStream::Propagate_Messages() Line 177	C++
```

```
Exception thrown: read access violation.
draw was nullptr.
```

```c++
            Drawable *draw = g_theInGameUI->Get_First_Selected_Drawable();

            for (auto it = m_idleWorkerLists[index].begin(); it != m_idleWorkerLists[index].end(); it++) {
                if (*it == draw->Get_Object()) {
                    if (it != m_idleWorkerLists[index].end()) {
                        obj = *it;
                    } else {
                        obj = *m_idleWorkerLists[index].begin();
                    }

                    break;
                }
            }
```

**Comment by xezon** [comment]
Perhaps forgot to hook constructor of InGameUI?

**Comment by xezon** [comment]
Selecting faction building and pressing E button will attempt selecting matching units. Is introduced with this change. Is not fixed by commenting all new hooks.

shot_20240201_182931_1

**Comment by xezon** [comment]
USA Sentry Drone + USA Humvee cannot attack together.

Issue appears to be introduced by `InGameUI::Get_Can_Selected_Objects_Attack`.

EDIT: Also happens in Original. Is related to order of construction. If Sentry Drone is constructed first, then Humvee cannot attack in group selection.

shot_20240201_191422_1

**Comment by jonwil** [comment]
Fixed idle worker crash.
Fixed issue with selecting a structure and pressing E.
InGameUI constructor is hooked in via W3DGameClient::Create_InGameUI

**Review by xezon — COMMENTED** [review]
Select next idle worker does not work. It will only select the first worker on every button press.

shot_20240201_195707_1

**Comment by jonwil** [comment]
Fixed idle worker selection.

**Review by xezon — APPROVED** [review]
Select matching units on screen and map works
Region select and draw works
Floating text works
Draw move hint works
Draw place angle works
Group repair works
Group docking works
Group healing works
Group resume construction works
Group hijack works
Group enter works
Group enter bomb car works
Group disable vehicle works
Set Rally point works
Group combat drop works
Group sabotage building works
Special Power at target works
Special Power at location works
Group attack works
Hidden tool tip during military briefing works
Select next and previous idle worker works
Hide worker idle button in Replay works

**Pull request merged**